keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -288,9 +288,39 @@ like the `all` macro, e.g. `self.all(listItem, <predicate>)` or `self.all(mapKey
   - xref [analysis of possible interactions with immutability and
     validation](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1101-immutable-fields#openapi-extension-x-kubernetes-immutable).
 
-- If a object property collides with a CEL keyword (see RESERVED in [CEL Syntax](https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax)),
-  it will be escaped by prepending a _ prefix. To prevent this from causing a subsequent collision, all properties with a `_` prefix will always be
+- If a object property name is a CEL keyword (see RESERVED in [CEL Syntax](https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax)),
+  it will be escaped by prepending a _ prefix. To prevent this from causing a subsequent collision,  properties named with a CEL keyword and a  `_` prefix will be
   prefixed by `__` (generally, N+1 the existing number of `_`s).
+  
+- If a object property name is a CEL language identifier (`int`, `uint`, `double`, `bool`, `string`,
+  `bytes`, `list`, `map`, `null_type`, `type`, see [CEL language
+  identifiers](https://github.com/google/cel-spec/blob/master/doc/langdef.md#values)) it is not
+  accessible as a root variable and must be accessed via `self`, .e.g. `self.int`.
+
+- Only property names of the form `[a-zA-Z_.-/][a-zA-Z0-9_.-/]*` are accessible.
+  If a property name is "self" or matches with a [reserved language identifier](https://github.com/google/cel-spec/blob/v0.6.0/doc/langdef.md#values) 
+  (`int`, `uint`, `double`, `bool`, `string`, `bytes`, `list`, `map`, `null_type`, `type`), it is
+  not escaped, but it is excluded from the bound variables and can only be accessed via 
+  "self.{property name}". All other accessible property names are escaped according to the following rules
+  when accessed in the expression:
+	- '__' escapes to '__underscores__'
+	- '.' escapes to '__dot__'
+	- '-' escapes to '__dash__'
+	- '/' escapes to '__slash__'
+	- CEL RESERVED keywords escape to '__{keyword}__'. The keywords are: "true", "false", "null",
+	"in", "as", "break", "const", "continue", "else", "for", "function", "if", "import", "let",
+	"loop", "package", "namespace", "return".
+
+- Rules may be written at the root of an object, and may make field selection into any fields
+  declared in the OpenAPIv3 schema of the CRD as well as `apiVersion`, `kind`, `metadata.name` and
+  `metadata.generateName`. This includes selection of fields in both the `spec` and `status` in the
+  same expression, e.g. `status.quantity <= spec.maxQuantity`. Because CRDs only allow the `name`
+  and `generatName` to be declared in the `metadata` of an object, these are the only metadata
+  fields that may be validated using CEL validator rules. For example,
+  `metadata.name.endsWith('mySuffix')` is allowed, but `size(metadata.labels) < 3` it not
```

Comment:
nit: is not

Reply:
Can force something like "must have a label called foo" or "if metadata.labels["foo"] exists, it must be one of 'alfa', 'bravo'"?

Reply:
Only 'name' and 'generateName' are accessible to OpenAPIv3 restrictions. xref: rule 4 from https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema. Per discussions with @sttts we are extending that restriction to CEL validator rules as well.

For expressing the rules:

- "must have a map key called foo" would be `'foo' in mapField`.
- "if mapField["foo"] exists, it must be one of 'alfa', 'bravo'" would be `!('foo' in mapField) || mapField['foo'] in ['alfa', 'bravo']`